File: src/background/base-url.test.ts
import { describe, expect, it } from "vitest";
import {
  chatCompletionsUrl,
  hostPermissionPattern,
  isLoopbackBaseUrl,
  normalizeBaseUrl,
} from "./base-url";

describe("model profile URL safety", () => {
  it("normalizes a trailing slash from an HTTPS base URL", () => {
    expect(normalizeBaseUrl("https://api.deepseek.com/v1/")).toBe("https://api.deepseek.com/v1");
  });

  it("appends the chat completions endpoint", () => {
    expect(chatCompletionsUrl("https://api.deepseek.com/v1")).toBe(
      "https://api.deepseek.com/v1/chat/completions",
    );
  });

  it("does not append a duplicate chat completions endpoint", () => {
    expect(chatCompletionsUrl("http://localhost:11434/v1/chat/completions")).toBe(
      "http://localhost:11434/v1/chat/completions",
    );
  });

  it.each(["http://localhost:11434/v1", "http://127.0.0.1:11434/v1"])(
    "allows a local HTTP model endpoint at %s",
    (baseUrl) => {
      expect(normalizeBaseUrl(baseUrl)).toBe(baseUrl);
    },
  );

  it("rejects remote HTTP model endpoints", () => {
    expect(() => normalizeBaseUrl("http://api.example.com/v1")).toThrow("HTTPS");
  });

  it("rejects embedded URL credentials", () => {
    expect(() => normalizeBaseUrl("https://user:pass@example.com/v1")).toThrow("credentials");
  });

  it.each([
    "https://api.example.com/v1?",
    "https://api.example.com/v1/chat/completions?",
    "https://api.example.com/v1?tenant=syntax",
    "https://api.example.com/v1/chat/completions?tenant=syntax",
  ])("rejects a query string in model URL %s", (baseUrl) => {
    expect(() => normalizeBaseUrl(baseUrl)).toThrow("query strings or fragments");
    expect(() => chatCompletionsUrl(baseUrl)).toThrow("query strings or fragments");
  });

  it.each([
    "https://api.example.com/v1#",
    "https://api.example.com/v1/chat/completions#",
    "https://api.example.com/v1#syntax",
    "https://api.example.com/v1/chat/completions#syntax",
  ])("rejects a fragment in model URL %s", (baseUrl) => {
    expect(() => normalizeBaseUrl(baseUrl)).toThrow("query strings or fragments");
    expect(() => chatCompletionsUrl(baseUrl)).toThrow("query strings or fragments");
  });

  it("derives an exact-origin host permission", () => {
    expect(hostPermissionPattern("https://api.deepseek.com:8443/v1")).toBe(
      "https://api.deepseek.com:8443/*",
    );
  });
});

/**
 * 本地与云端的合批取舍相反:本地模型串行处理请求,合并成一条大请求才快
 * (CHANGELOG 1.0.4 记录的收益);云端 API 并发好,而耗时几乎只由输出 token 决定,
 * 拆小批并行才快。所以要能区分二者。
 */
describe("isLoopbackBaseUrl", () => {
  it.each([
    ["http://localhost:11434/v1", true],
    ["http://127.0.0.1:1234/v1", true],
    ["https://api.deepseek.com", false],
    ["https://api.openai.com/v1", false],
  ])("%s → %s", (baseUrl, expected) => {
    expect(isLoopbackBaseUrl(baseUrl)).toBe(expected);
  });

  it("对畸形 URL 返回 false 而不是抛错——判定失败该退回云端策略", () => {
    expect(isLoopbackBaseUrl("not a url")).toBe(false);
  });
});

File: intellij-plugin/src/test/kotlin/dev/codetui/englishsyntax/model/BaseUrlTest.kt
package dev.codetui.englishsyntax.model

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class BaseUrlTest {
  @Test
  fun `normalizes a trailing slash from an HTTPS base URL`() {
    assertEquals("https://api.deepseek.com/v1", normalizeBaseUrl("https://api.deepseek.com/v1/"))
  }

  @Test
  fun `appends the chat completions endpoint`() {
    assertEquals(
      "https://api.deepseek.com/v1/chat/completions",
      chatCompletionsUrl("https://api.deepseek.com/v1"),
    )
  }

  @Test
  fun `does not append a duplicate chat completions endpoint`() {
    assertEquals(
      "http://localhost:11434/v1/chat/completions",
      chatCompletionsUrl("http://localhost:11434/v1/chat/completions"),
    )
  }

  @Test
  fun `allows a local HTTP model endpoint`() {
    assertEquals("http://localhost:11434/v1", normalizeBaseUrl("http://localhost:11434/v1"))
    assertEquals("http://127.0.0.1:11434/v1", normalizeBaseUrl("http://127.0.0.1:11434/v1"))
  }

  @Test
  fun `rejects remote HTTP model endpoints`() {
    val error = assertFailsWith<IllegalArgumentException> {
      normalizeBaseUrl("http://api.example.com/v1")
    }
    assertTrue(error.message.orEmpty().contains("HTTPS"))
  }

  @Test
  fun `rejects embedded URL credentials`() {
    val error = assertFailsWith<IllegalArgumentException> {
      normalizeBaseUrl("https://user:pass@example.com/v1")
    }
    assertTrue(error.message.orEmpty().contains("credentials"))
  }

  @Test
  fun `rejects query strings and fragments`() {
    listOf(
      "https://api.example.com/v1?",
      "https://api.example.com/v1?tenant=syntax",
      "https://api.example.com/v1#",
      "https://api.example.com/v1#syntax",
    ).forEach { baseUrl ->
      assertFailsWith<IllegalArgumentException>(baseUrl) { normalizeBaseUrl(baseUrl) }
      assertFailsWith<IllegalArgumentException>(baseUrl) { chatCompletionsUrl(baseUrl) }
    }
  }

  @Test
  fun `normalizes scheme and host case and strips default ports`() {
    assertEquals("https://api.example.com/v1", normalizeBaseUrl("HTTPS://API.EXAMPLE.COM/v1"))
    assertEquals("https://api.example.com/v1", normalizeBaseUrl("https://api.example.com:443/v1"))
    assertEquals("http://localhost/v1", normalizeBaseUrl("http://localhost:80/v1"))
  }

  @Test
  fun `rejects a URL without a host`() {
    assertFailsWith<IllegalArgumentException> { normalizeBaseUrl("https:///path") }
  }

  @Test
  fun `isLoopbackBaseUrl distinguishes local from remote and tolerates malformed input`() {
    assertTrue(isLoopbackBaseUrl("http://localhost:11434/v1"))
    assertTrue(isLoopbackBaseUrl("http://127.0.0.1:1234/v1"))
    assertFalse(isLoopbackBaseUrl("https://api.deepseek.com"))
    assertFalse(isLoopbackBaseUrl("https://api.openai.com/v1"))
    assertFalse(isLoopbackBaseUrl("not a url"))
  }
}
